scipy/stats/_distn_infrastructure.py

```diff
@@ -2853,6 +2855,128 @@ def entropy(pk, qk=None, base=None, axis=0):
     return S
 
 
+def differential_entropy(
+    values: np.ndarray,
+    *,
+    window_length: Optional[int] = None,
+    base: Optional[float] = None,
+    axis: int = 0,
+) -> np.ndarray:
+    r"""Given a sample of a distribution, calculate the differential entropy.
+
+    This routine uses the Vasicek estimator of the differential entropy. Given
+    a sorted random sample :math:`X_1, \ldots X_n` this is defined as:
+
+    .. math::
+        \frac{1}{n}\sum_1^n \log\left[ \frac{n}{2m} (X_{i+m} - X_{i-m}) \right]
+
+    where :math:`m` is the window length parameter, :math:`X_{i} = X_1` if
+    :math:`i < 1` and :math:`X_{i} = X_n` if :math:`i > n`.
+
+    Parameters
+    ----------
+    values : sequence
+        Samples of the (continuous) distribution.
+    window_length : int, optional
+        Window length for computing Vasicek estimate. Must be an integer
+        between 1 and half of the sample size. If ``None`` (the default) it
+        uses the heuristic value
+
+        .. math::
+            \left \lfloor \sqrt{n} + 0.5 \right \rfloor
+
+        where :math:`n` is the sample size. This heuristic was originally
+        proposed in [2]_ and has become common in the literature.
+    base : float, optional
+        The logarithmic base to use, defaults to ``e`` (natural logarithm).
+    axis : int, optional
+        The axis along which the differential entropy is calculated.
+        Default is 0.
+
+    Returns
+    -------
+    entropy : float
+        The calculated differential entropy.
+
+    Notes
+    -----
+    This function will converge to the true differential entropy in the limit
+    when
+
+    .. math::
+        n \to \infty, \quad m \to \infty, \quad \frac{m}{n} \to 0
+
+    The optimal choice of ``window_length`` for a given sample size, depends on
+    the (unknown) distribution. In general, the smoother the density of the
+    distribution, the larger is such optimal value of ``window_length`` [1]_.
+
+    References
+    ----------
+    .. [1] Vasicek, O. (1976). A test for normality based on sample entropy.
+           Journal of the Royal Statistical Society:
+           Series B (Methodological), 38(1), 54-59.
+    .. [2] Crzcgorzewski, P., & Wirczorkowski, R. (1999). Entropy-based
+           goodness-of-fit test for exponentiality. Communications in
+           Statistics-Theory and Methods, 28(5), 1183-1202.
+
+    Examples
+    --------
+
+    >>> from scipy.stats import differential_entropy, norm
+
+    Entropy of a standard normal distribution:
+
+    >>> random_state = np.random.RandomState(0)
```

Comment:
I would not promote usage of `RandomState` one should use `Generator` for new code. But ok in tests as we cannot use yet generators here.

Reply:
I think that `Generator` has no guarantees of not changing the random stream in future updates. I would rather use `RandomState` for tests and reproducible results.

Reply:
For tests yes we have to stick with `RandomState` because we need fixed values between various envs and versions where `Generator` is not present yet. But NumPy's doc is clearly stating that new code should use `Generator`. So It's not good if in our examples we show old/bad practices.

Reply:
To be clear, I am talking about this: https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator

So you can fix the seed in the same way. This is the new way to draw random numbers from NumPy so this is actually more future proof. `RandomState` is there for compatibility reasons only now.

Reply:
Comments like "The default is currently PCG64 but this may change in future versions. " suggest me that one should not count on the particular random values of the random generator to be the same between numpy versions for a `Generator` object. Also in https://numpy.org/neps/nep-0019-rng-policy.html#supporting-unit-tests it seems to say that one should still use `RandomState` for reproducibility between numpy versions.

Reply:
Not blocking this.

But so I get this right for future reviews, @rgommers do we have guidelines for doctests? To me, we are exposing code to user, which many will just copy paste as usual. Hence the need to have clean examples and use as much as we can best practices.

Reply:
> But if `Generator` is changed by some future update, wouldn't doctests fail? Is this something that we do not care about?

This is possible but unlikely to happen in practice. And it's a price worth paying, we should use the new API (I'd use `np.random.default_rng(seed=148632)` here).

> it seems to say that one should still use `RandomState` for reproducibility between numpy versions.

This is true, but the only place you'd want that exact reproducibility is in the test suite. Documentation examples are meant to demonstrate best practices, they're not tests.

Reply:
Thanks for the precisions @rgommers. I've seen a few places where we advertise `RandomState` (such as stats tutorials), I will see to make a PR to direct users to use `Generator` here as well.

Reply:
> This is possible but unlikely to happen in practice. And it's a price worth paying, we should use the new API (I'd use `np.random.default_rng(seed=148632)` here).

Why that particular seed?

Reply:
Just some random larger number. I think this is still not perfect, but using `seed=0` is very low-entropy and if one gets that habit and does it everywhere that's not good. There's a long discussion somewhere, and I think in cases it matters the recommendation is to use something like a number high in the range of the span of the integer dtype you're using, so every bit in the 64-bit int you're using has a 50/50 chance of being 0 or 1.

Reply:
Maybe you are referring to https://numpy.org/doc/stable/reference/random/bit_generators/index.html#seeding-and-entropy. I was not aware that the seed you used have any impact in the quality of the random sequence. If that is the case, this is a little bit worrying, as most people I know uses 0 or 1 for the seeds of reproducible experiments. I will use your seed, with an appropriate comment.

Reply:
nice, yes that's it - didn't realize we now actually have good docs in NumPy for this:)

Reply:
Ok, doctest fixed. Let me know if it is ok now.

Reply:
LGTM!

Reply:
Here'acomment from Kevin Sheppard w.r.t. long integers for seeds:

https://github.com/numpy/numpy/pull/18014#discussion_r545165413

ISTM, four- or five- digit integer seeds are OK in scipy examples